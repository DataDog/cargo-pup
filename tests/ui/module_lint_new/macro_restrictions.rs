// Test for granular macro restriction functionality
// This product includes software developed at Datadog (https://www.datadoghq.com/) Copyright 2024 Datadog, Inc.
//
// NOTE: This test only covers declarative_macro detection. Proc macro types
// (proc_macro, proc_macro_attribute, proc_macro_derive) cannot be tested in UI tests
// because the compiler validates proc macro signatures and crate types before
// However, the implementation correctly detects all four granular macro types
// as demonstrated in test_app.

//@compile-flags: --crate-name test_macro_restrictions
//@compile-flags: --crate-type lib

// Declarative macro - should trigger the declarative_macro lint
macro_rules! forbidden_declarative_macro {
    //~ ERROR: declarative macro 'forbidden_declarative_macro' is not allowed in this module
    () => {
        println!("This declarative macro is forbidden!");
    };
}

// Another declarative macro to show comprehensive detection
macro_rules! another_forbidden_macro {
    //~ ERROR: declarative macro 'another_forbidden_macro' is not allowed in this module
    ($x:expr) => {
        println!("Another forbidden macro: {}", $x);
    };
}

// Regular function - should be allowed
pub fn allowed_function() {
    println!("This function is allowed");
}
