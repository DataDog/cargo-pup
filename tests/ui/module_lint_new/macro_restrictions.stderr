error: declarative macro 'forbidden_declarative_macro' is not allowed in this module
  --> tests/ui/module_lint_new/macro_restrictions.rs:12:1
   |
LL | / macro_rules! forbidden_declarative_macro {
LL | |     () => {
LL | |         println!("This declarative macro is forbidden!");
LL | |     };
LL | | }
   | |_^
   |
   = help: Consider moving this item to a different module
   = note: Applied by cargo-pup rule 'no_declarative_macros_allowed'.
   = note: `#[deny(module_denied_items)]` on by default

error: declarative macro 'another_forbidden_macro' is not allowed in this module
  --> tests/ui/module_lint_new/macro_restrictions.rs:19:1
   |
LL | / macro_rules! another_forbidden_macro {
LL | |     ($x:expr) => {
LL | |         println!("Another forbidden macro: {}", $x);
LL | |     };
LL | | }
   | |_^
   |
   = help: Consider moving this item to a different module
   = note: Applied by cargo-pup rule 'no_declarative_macros_allowed'.

error: aborting due to 2 previous errors

